### PR TITLE
Sound Expanded: Add Sound Expanded extension

### PR DESCRIPTION
**Pull request by LilyMakesThings**
Sound Expanded is an extension that adds a few missing blocks from the Sound category.

This serves as a precursor to my next extension, Audio Engine.
image

### CODE REVIEW

**Comment by NexusKitten** [comment]
sound-category booleans are cursed

**Comment by AdamMady** [comment]
this reminds me of when I tried to make a "while playing note {VAR} for {VAR} beats in the {VAR}th beat fade into note {VAR} for {VAR} beats" block (it went badly)

**Comment by LilyMakesThings** [comment]
> isnt start sound on loop the same as start looping sound?

"start looping ()" only works for sounds that are already playing.

**Comment by AdamMady** [comment]
I have a good idea @LilyMakesThings how about you make a block where you could change the global volume of the project instead of changing each volume of each sprite one by one

**Comment by AdamMady** [comment]
I think I'll have a go for both I have nothing better to do

**Comment by BlueDome77** [comment]
Could you add these blocks?

scratchblocks-2

**Comment by LilyMakesThings** [comment]
Please look at the block palette before suggesting blocks

**Comment by AdamMady** [comment]
I know this is a bit much for a simple extension but is it possible to do a record block? where it records using the microphone and saves it to be used

**Comment by BlueDome77** [comment]
> I know this is a bit much for a simple extension but is it possible to do a record block? where it records using the microphone and saves it to be used

There is an extension exactly for that! Find "Recording" in the list: https://github.com/BlueDome77/Turbowarp-Extension-List/blob/main/README.md

**Comment by BlueDome77** [comment]
I think you should merge the Sound extension with this

**Comment by BlueDome77** [comment]
Could you add these?

scratchblocks-2

**Comment by LilyMakesThings** [comment]
> Could you add these?
> 
> scratchblocks-2

No

**Comment by Martinelplayz** [comment]
> Could you add these?
> 
> scratchblocks-2

:c

**Comment by NexusKitten** [comment]
can't wait!

File: extensions/Lily/SoundExpanded.js [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+      return -1;
+    }
+
+    _wrapClamp(n, min, max) {
+      const range = (max - min) + 1;
+      return n - (Math.floor((n - min) / range) * range);
+    }
+
+    _getTargets() {
+      const spriteNames = [];
```
Comment (GarboMuffin) [inline extensions/Lily/SoundExpanded.js:323]: let's avoid the same mistake as looks+. have the top value always be "myself" with value `_myself_` and just list every sprite under it, and also add the stage (`_stage_`)

Comment (LilyMakesThings) [reply extensions/Lily/SoundExpanded.js:323]: I made a script that does that for skins, I'll add that now

**Comment by GarboMuffin** [comment]
the combination of "start sound [ ] on loop" and "start looping [ ]" is going to confuse many people

**Comment by LilyMakesThings** [comment]
> the combination of "start sound [ ] on loop" and "start looping [ ]" is going to confuse many people

If I add "while it's playing" or something to the end, would that work? or could I literally just remove it

**Comment by CST1229** [comment]
> If I add "while it's playing" or something to the end, would that work? or could I literally just remove it

maybe `begin/end looping for []`

**Comment by LilyMakesThings** [comment]
> > If I add "while it's playing" or something to the end, would that work? or could I literally just remove it
> 
> maybe `begin`/`end looping for []`

That's the one I think

**Comment by LilyMakesThings** [comment]
Update: added project volume blocks. Works (mostly) with addons' volume slider.
image

**Comment by LilyMakesThings** [comment]
I added something an attribute for dataURIs so you can export the sound as a data uri.

I pretty much only did this for the Sound extension, as well as the Video extension.

File: extensions/Lily/SoundExpanded.js [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                type: Scratch.ArgumentType.SOUND,
+              },
+              START: {
+                type: Scratch.ArgumentType.NUMBER,
+                defaultValue: 0,
+              },
+            },
+          },
+          {
+            opcode: "stopLooping",
```
Comment (GarboMuffin) [inline extensions/Lily/SoundExpanded.js:45]: is it intended that this doesn't actually stop the sound immediately?

File: extensions/Lily/SoundExpanded.js [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    setProjectVolume(args) {
+      const value = Scratch.Cast.toNumber(args.VALUE);
+      const newVolume = this._wrapClamp(value / 100, 0, 1);
+      runtime.audioEngine.inputNode.gain.value = newVolume;
+    }
+
+    changeProjectVolume(args) {
+      const value = Scratch.Cast.toNumber(args.VALUE) / 100;
+      const volume = runtime.audioEngine.inputNode.gain.value;
+      const newVolume = this._wrapClamp(volume + value, 0, 1);
```
Comment (GarboMuffin) [inline extensions/Lily/SoundExpanded.js:309]: when volume is 0 and someone changes it by -1, it probably shouldn't wrap around to 199%

**Comment by GarboMuffin** [comment]
Also how does this compare to https://github.com/TurboWarp/extensions/pull/756?

File: extensions/Lily/SoundExpanded.js (hunk 2 above)
Comment (LilyMakesThings) [reply extensions/Lily/SoundExpanded.js:32]: yes, that's why there's a "stop sound" block.

**Comment by LilyMakesThings** [comment]
> Also how does this compare to #756?

ultrasound appears to not work in line with the default sound category, rather on its own use of the web audio api - think audio engine, audiostream or tuneshark. the main purpose of this is no different to clones plus, looks plus, etc.

File: extensions/Lily/SoundExpanded.js (hunk 3 above)
Comment (LilyMakesThings) [reply extensions/Lily/SoundExpanded.js:296]: Good point, will change that

**Comment by GarboMuffin** [comment]
done?

**Comment by LilyMakesThings** [comment]
yes

**Pull request merged**